pkg/services/ngalert/api/api_alertmanager_guards.go

```diff
@@ -0,0 +1,148 @@
+package api
+
+import (
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/util/cmputil"
+	"github.com/prometheus/alertmanager/pkg/labels"
+)
+
+func (srv AlertmanagerSrv) provenanceGuard(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	if err := checkRoutes(currentConfig, newConfig); err != nil {
+		return err
+	}
+	if err := checkTemplates(currentConfig, newConfig); err != nil {
+		return err
+	}
+	if err := checkContactPoints(currentConfig.AlertmanagerConfig.Receivers, newConfig.AlertmanagerConfig.Receivers); err != nil {
+		return err
+	}
+	if err := checkMuteTimes(currentConfig, newConfig); err != nil {
+		return err
+	}
+	return nil
+}
+
+func checkRoutes(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	reporter := cmputil.DiffReporter{}
+	ops := []cmp.Option{cmp.Reporter(&reporter), cmpopts.EquateEmpty(), cmpopts.IgnoreUnexported(labels.Matcher{})}
+	routesEqual := cmp.Equal(currentConfig.AlertmanagerConfig.Route, newConfig.AlertmanagerConfig.Route, ops...)
+	if !routesEqual && currentConfig.AlertmanagerConfig.Route.Provenance != ngmodels.ProvenanceNone {
+		return fmt.Errorf("policies were provisioned and cannot be changed through the UI")
+	}
+	return nil
+}
+
+func checkTemplates(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	for name, template := range currentConfig.TemplateFiles {
+		found := false
+		provenance := ngmodels.ProvenanceNone
+		if prov, present := currentConfig.TemplateFileProvenances[name]; present {
+			provenance = prov
+		}
```

Comment:
I think we can simplify some of the logic here by elevating the check for `provenance` that currently happens in several places. As in,

```
for name, template := range currentConfig.TemplateFiles {
    provenance := <get provenance for current template>
    if provenance == ngmodels.ProvenanceNone {
        break
    }
    <rest of logic>
}
```

Idea being, we don't care about guarding non-provisioned objects. So then, the rest of the code can assume `provenance != ngmodels.ProvenanceNone` which should simplify things.